### ISSUE TITLE
ProgressCallback should be throttled instead of debounced

### ISSUE THREAD

**Issue by jedwards1211** [microsoft/vscode#86131]
This explains the cause of #54295.

The progress callback is decorated with `@debounced`:

https://github.com/microsoft/vscode/blob/d57b8538763b9db16bd8ca3b4dd17f5aa129b87f/src/vs/workbench/api/common/extHostProgress.ts#L88

This basically means that while I'm continuously reporting progress updates, the displayed progress doesn't update.  It only updates once it's been more than 100 milliseconds since I last reported progress.

I think what you meant to do is *throttle* the progress updates (update the displayed progress once every 100 milliseconds) rather than *debounce* (only update once it's been 100 milliseconds since the last call).

Basically to work around this I have to throttle my own calls to `progress.report` to > 100 milliseconds.

@bpasero also why roll your own debounce implementation instead of just using `lodash` debounce/throttle?  If you were using `lodash` I could easily PR a fix to use `lodash`'s `throttle` instead.

**Comment by jedwards1211** [comment]
The OP on #54295 had presumed the following:
> It "merges" repeated progress calls into a single progress RPM, but a faulty conditional prevents repeated increment values from being added together to sum up to the final value.

However AFAICT the increment reducing works fine, the problem is just that updates are debounced instead of throttled.

**Comment by bpasero** [comment]
@jedwards1211 yeah I think you are on the right path, can you come up with a sample extension that demonstrates the problem simply by producing progress updates in a way that it shows the issue and proofs your fix?

**Comment by jedwards1211** [comment]
Sure, I may not get to that until Monday but will do!

**Comment by jedwards1211** [comment]
@bpasero okay, I created https://github.com/jedwards1211/progress-issue-demo to demonstrate the issue.  Instructions are in the README
